### DISCUSSION TITLE
Uploading Firmware to the Nano using Mac OS Big Sur With avrdude

### DISCUSSION THREAD

**Discussion by ic-racer** [srozum/film_camera_tester#33] [Q&A] [answered]
Some thoughts for discussion on how to do this on mac...I have not tried it yet.

1) The flash uploading software 'avrdude' gets installed on one's Mac when installing 'Arduino IDE' so, no need to install it as a stand alone (see screenshot of 'Terminal' screen from command **avrdude -?** below to confirm 'avrdude' is actually installed and active.)
2) To see how 'avrdude' is used to upload a .hex, in 'Arduino IDE 2.0.4' preferences, check 'Verbose' then...
3) Connect and upload 'blink' to a Nano in 'Arduino IDE' , _just to see the verbose output_. This output will have the information to 'fill in the blanks' when writing the parameters to send to avrdude

4) Upload "ArduinoISP" sketch to a designated uploader board
5) Connect the board on which one wants 'firmware.hex' to the uploader board that has "ArduinoISP" on it. 

<img width="795" alt="Screen Shot 2023-04-17 at 5 46 25 PM" src="https://user-images.githubusercontent.com/128929440/232618034-6ad3ac7f-2a8c-4813-81cf-38c00b552c65.png">

**Comment by ic-racer** [comment]
Since the verbose output of Arduino IDE gives the path to the 'blink' hex file, I figured that would be a convenient location to place the 'firmware.hex' file. Because that way I can just copy the complex path. 

<img width="674" alt="Screen Shot 2023-04-17 at 5 11 34 PM" src="https://user-images.githubusercontent.com/128929440/232619729-b22fd8f4-fff6-4964-9507-83a7181be3ee.png">

**Comment by ic-racer** [comment]
I copied the verbose output parameters from Arudino IDE to see how I needed to control avrdude. Then I modified them in red to work with the ArduinoISP board being used as the uploader. (-cavrisp).

After erase and unlock board, then set fuses, avrdude is directed to the firmware.hex file in the same directory where blink resides. 

Still working on this, so I have not yet added the commands to upload the default settings to eprom. 

<img width="1136" alt="Screen Shot 2023-04-17 at 5 44 42 PM" src="https://user-images.githubusercontent.com/128929440/232620422-2d16c121-13d6-422c-bc22-f581f22acb7d.png">

**Comment by ic-racer** [comment]
I ran the three avrdude program 3 times to 1) unlock the board, 2) set the fuses, and 3) write 'firmware.hex' and settings.hex.'

Seemed to go OK but looking at the output **it indicated an error setting fuses and they were all set to "00"**
even though it indicated that 'firware.hex' and 'settings.hex' were properly loaded. 

I went back and tried to unlock the board again and all I get when I try "-Ulock:w:0x3F:m" is :

avrdude: stk500_recv(): programmer is not responding
avrdude: stk500_getsync() attempt 1 of 10: not in sync: resp=0x00
avrdude: stk500_getsync() attempt 2 of 10: not in sync: resp=0x1c
avrdude: stk500_getsync() attempt 3 of 10: not in sync: resp=0x1c
avrdude: stk500_getsync() attempt 4 of 10: not in sync: resp=0x1c
avrdude: stk500_getsync() attempt 5 of 10: not in sync: resp=0x1c
avrdude: stk500_getsync() attempt 6 of 10: not in sync: resp=0x1c
avrdude: stk500_getsync() attempt 7 of 10: not in sync: resp=0x1c
avrdude: stk500_getsync() attempt 8 of 10: not in sync: resp=0x1c
avrdude: stk500_getsync() attempt 9 of 10: not in sync: resp=0x1c
avrdude: stk500_getsync() attempt 10 of 10: not in sync: resp=0x1c

avrdude done.  Thank you.

I tried to reload the 'ArduinoISP' to all three of my Nano boards and it seems to load ok, but when I try the -Ulock:w:0x3F:m command in avrdude, all I get is the above error. 

I have two more boards that just came in the mail. Will try those as the programmer and see what happens.

**Comment by srozum** [comment]
First of all, I don't understand why you uploading firmware blindly without assembled hardware to test it.

Second, if you was able to upload 'Arduino ISP' to all three of your boards via USB cable, means that all boards weren't unlocked nor erased when you ran avrdude commands which change fuses. Note, that you cannon change fuses on a board connected via USB, only Arduino as ISP can change fuses on TARGET board.

Commands you posted above seems fine, same as mine but with path added. Means you doing something wrong, and I cannot say what exactly with given information.

**Comment by ic-racer** [comment]
Thank you for the response. So, now I know the Arduino ISP is needed to set fuses. 
So now, the fuse settings are the issue. I did experiment uploading the .hex files via usb (no two board connection) and got this output below. It appears to indicate the .hex files were successfully loaded but shows the fuses are 00. I suspect that board not going to work because of that?

Will go back and try to get the ISP uploader to work again. It (Arduino ISP uploader) did work only once but after that it stopped responding. Don't know why. Will continue to experiment. 

Hardware is slow to build due to prolonged wait for the parts. So far almost 2 months. The LED was horribly non-functional (see below) so I ordered more from 2 different suppliers to get a good one. The rotary selector came without the two switch leads, so that needed to be re-ordered also. Nothing too serious, still moving forward with the build.

OUTPUT FROM AVRDUDE:

```
avrdude: Version 6.3-20190619
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch
```
```

         System wide configuration file is "/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/etc/avrdude.conf"
         User configuration file is "/Users/daddy/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping

         Using Port                    : /dev/cu.usbserial-14110
         Using Programmer              : arduino
         Overriding Baud Rate          : 57600
         AVR Part                      : ATmega328P
         Chip Erase delay              : 9000 us
         PAGEL                         : PD7
         BS2                           : PC2
         RESET disposition             : dedicated
         RETRY pulse                   : SCK
         serial program mode           : yes
         parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         ByteDelay                     : 0
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                  Block Poll               Page                       Polled
           Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom        65    20     4    0 no       1024    4      0  3600  3600 0xff 0xff
           flash         65     6   128    0 yes     32768  128    256  4500  4500 0xff 0xff
           lfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           hfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           efuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           lock           0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           calibration    0     0     0    0 no          1    0      0     0     0 0x00 0x00
           signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00

         Programmer Type : Arduino
         Description     : Arduino
         Hardware Version: 2
         Firmware Version: 1.16
         Vtarget         : 0.0 V
         Varef           : 0.0 V
         Oscillator      : Off
         SCK period      : 0.1 us

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e950f (probably m328p)
avrdude: safemode: lfuse reads as 0
avrdude: safemode: hfuse reads as 0
avrdude: safemode: efuse reads as 0
avrdude: reading input file "/private/var/folders/gb/fqk5qm6d5s9057zzjf_vhc5w0000gn/T/arduino/sketches/firmware.hex"
avrdude: writing flash (32126 bytes):

Writing | ################################################## | 100% 10.29s

avrdude: 32126 bytes of flash written
avrdude: reading input file "/private/var/folders/gb/fqk5qm6d5s9057zzjf_vhc5w0000gn/T/arduino/sketches/settings.hex"
avrdude: writing eeprom (1024 bytes):

Writing | ################################################## | 100% 4.90s

avrdude: 1024 bytes of eeprom written

avrdude: safemode: lfuse reads as 0
avrdude: safemode: hfuse reads as 0
avrdude: safemode: efuse reads as 0
avrdude: safemode: Fuses OK (E:00, H:00, L:00)

avrdude done.  Thank you.
```

![faulty led](https://user-images.githubusercontent.com/128929440/233515923-8280afc8-239c-42a4-9118-652c5526748d.jpg)

**Comment by ic-racer** [comment]
I retrieved the Terminal log and found out the first time I used the -Ulfuse:w:0xFF:m -Uhfuse:w:0xD7:m -Uefuse:w:0xFF:m commands with avrdude and my ISP Programmer I got the following reply:

```
avrdude: safemode: lfuse reads as 0
avrdude: safemode: hfuse reads as 0
avrdude: safemode: efuse reads as 0
avrdude: safemode: lfuse changed! Was ff, and is now 0
Would you like this fuse to be changed back? [y/n] n
avrdude: safemode: hfuse changed! Was d7, and is now 0
Would you like this fuse to be changed back? [y/n] n
avrdude: safemode: efuse changed! Was ff, and is now 0
Would you like this fuse to be changed back? [y/n] n
avrdude: safemode: Fuses OK (E:FF, H:D7, L:FF)
avrdude: stk500_recv(): programmer is not responding
```
Thereafter, every time I try to use my ISP Programmer board with avrdude I only get:
`avrdude: stk500_recv(): programmer is not responding`

**Comment by ic-racer** [comment]
To test my ISP Programmer, I opened the Arduino IDE program used it to "Update Bootloader" on a board and I got the following response, which seems to indicate that it worked and my ISP Programmer and its connections are OK. 

I also checked the instructions to avrdude passed from Arduino IDE program for the fuse setting commands and found it is using "-cstk500v1" instead of "-carduino" so the next thing to try is to use "-cstk500v1" with my ISP Programmer. 


```
avrdude: Version 6.3-20190619
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/etc/avrdude.conf"
         User configuration file is "/Users/daddy/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping

         Using Port                    : /dev/cu.usbserial-14110
         Using Programmer              : stk500v1
         Overriding Baud Rate          : 19200
         AVR Part                      : ATmega328P
         Chip Erase delay              : 9000 us
         PAGEL                         : PD7
         BS2                           : PC2
         RESET disposition             : dedicated
         RETRY pulse                   : SCK
         serial program mode           : yes
         parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         ByteDelay                     : 0
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                  Block Poll               Page                       Polled
           Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom        65    20     4    0 no       1024    4      0  3600  3600 0xff 0xff
           flash         65     6   128    0 yes     32768  128    256  4500  4500 0xff 0xff
           lfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           hfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           efuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           lock           0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           calibration    0     0     0    0 no          1    0      0     0     0 0x00 0x00
           signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00

         Programmer Type : STK500
         Description     : Atmel STK500 Version 1.x firmware
         Hardware Version: 2
         Firmware Version: 1.18
         Topcard         : Unknown
         Vtarget         : 0.0 V
         Varef           : 0.0 V
         Oscillator      : Off
         SCK period      : 0.1 us

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.02s

avrdude: Device signature = 0x1e950f (probably m328p)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file "/Users/daddy/Library/Arduino15/packages/arduino/hardware/avr/1.8.6/bootloaders/atmega/ATmegaBOOT_168_atmega328.hex"
avrdude: writing flash (32670 bytes):

Writing | ################################################## | 100% 0.00s

avrdude: 32670 bytes of flash written
avrdude: verifying flash memory against /Users/daddy/Library/Arduino15/packages/arduino/hardware/avr/1.8.6/bootloaders/atmega/ATmegaBOOT_168_atmega328.hex:
avrdude: load data flash data from input file /Users/daddy/Library/Arduino15/packages/arduino/hardware/avr/1.8.6/bootloaders/atmega/ATmegaBOOT_168_atmega328.hex:
avrdude: input file /Users/daddy/Library/Arduino15/packages/arduino/hardware/avr/1.8.6/bootloaders/atmega/ATmegaBOOT_168_atmega328.hex contains 32670 bytes
avrdude: reading on-chip flash data:

Reading | ################################################## | 100% 0.00s

avrdude: verifying ...
avrdude: 32670 bytes of flash verified
avrdude: reading input file "0x0F"
avrdude: writing lock (1 bytes):

Writing | ################################################## | 100% 0.02s

avrdude: 1 bytes of lock written
avrdude: verifying lock memory against 0x0F:
avrdude: load data lock data from input file 0x0F:
avrdude: input file 0x0F contains 1 bytes
avrdude: reading on-chip lock data:

Reading | ################################################## | 100% 0.01s

avrdude: verifying ...
avrdude: 1 bytes of lock verified

avrdude done.  Thank you.

```

**Reply by srozum** [reply]
Ok, few questions:

- except this log, can you provide the command what IDE was executing?
- why "Programmer Type : STK500" ?

Then, if you successfully burned bootloader, can you then connect that board via USB and upload simple Blink sketch? This will verify that bootloader was indeed burned successfully.

Also, explore avrdude attributes. You can READ fuses to know how your board is set. Like -U lock:r

**Comment by srozum** [comment]
And one more thing. Please, don't get me wrong. I am here to help with the build, but not to teach how to solder or program arduinos. Arduino is the simplest among the microcontrollers and was designed for educational purposes. There are tons of educational videos and articles on internet how to program arduinos.

You also can ask questions on https://forum.arduino.cc, it is great community always willing to help.

**Comment by ic-racer** [comment]
Ok, I can take the discussion over there. Rather than bothering you I was hoping another OSX user would chime it, realizing that I'm using UNIX version of avrdude and you instructions are windows based. 

I do want to get the burning sequence correct and document what did work before adding the steps to my "Build Thread"

Just to answer where "STK500" came from, this is the entire instruction that Arduino IDE sent to avrdude to burn the bootloader:

`/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude -C/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/etc/avrdude.conf -v -patmega328p -cstk500v1 -P/dev/cu.usbserial-14110 -b19200 -e -Ulock:w:0x3F:m -Uefuse:w:0xFD:m -Uhfuse:w:0xDA:m -Ulfuse:w:0xFF:m`

**Reply by srozum** [reply]
> realizing that I'm using UNIX version of avrdude and you instructions are windows based.

No, my instruction are for both platforms. I personally use Mac, and those are command which I used. And windows commands were tested and provided by another user who was upgrading his Tester.

The whole process was done by many people and proven to be correct.

> Just to answer where "STK500" came from, this is the entire instruction that Arduino IDE sent to avrdude to burn the bootloader:

This '-cstk500v1' part of a command is coming from "Programmer" setting in Arduino IDE. If you chose Arduino as ISP is should be '-cavrisp'. When uploading sketches normally via USB (using bootloader), you set programmer as "AVRISP mkII', which results in command as '-carduino'

**Reply by srozum** [reply]
Ah, I see where it comes from. There are few options to choose, "ArduinoISP" and "Arduino as ISP"

<img width="643" alt="Screenshot 2023-04-21 at 3 19 57 PM" src="https://user-images.githubusercontent.com/56254/233743048-9652e365-1643-437b-9de1-8e5f339bd6a4.png">

**Comment by srozum** [comment]
If you verified that your Arduino as ISP setup works, mean both boards connected correctly and avrdude can talk to them. If you are using Mac, only thing you have to do is run those commands from Wiki. Since you are using Arduino ISP instead of USBAsp programmer, you only have to replace 'usbasp' with 'avrisp' and port name, that's it.

'cd' to the folder where you downloaded firmware.hex file and run these commands:

avrdude -p atmega328p -c avrisp -P /dev/cu.usbserial-14110 -e -U lock:w:0x3F:m
avrdude -p atmega328p -c avrisp -P /dev/cu.usbserial-14110 -D -U lfuse:w:0xFF:m -U hfuse:w:0xD7:m -U efuse:w:0xFF:m
avrdude -p atmega328p -c avrisp -P /dev/cu.usbserial-14110  -D -U flash:w:firmware.hex:i -U eeprom:w:settings.hex:i

**Comment by ic-racer** [comment]
OK will try that. I was just using the long path to avrdude because that is what appears in Arduino IDE's verbose output. 

I'm pretty excited about the build two more packages came today. I have over 90% of the parts now.

**Comment by ic-racer** [comment]
OK, I see what might be an issue. I'm trying to run avrdude that was installed with Arduino IDE from the "Terminal" program in OSX. For example just typing "avrdude" yields the response "-bash: avrdude: command not found"

So, I'll try installing MacPorts and load avrdude from that and then use the commands recommended above.

**Reply by srozum** [reply]
No, adding path is fine. Avrdude will be the same if it's installed separately or with other IDE. That's wasn't a point.

Correct arguments like -c and -P and make sure that boards connected correctly. 

I have feeling that you are not totally understand what is supposed to be done. Please, read wiki part about why is programmer required: to set fuses, which disables bootloader, and upload code directly because bootloader is disabled.

**Comment by ic-racer** [comment, marked as answer]
Thank you for your help on this!!!  

 The baud rate command "-b57600" was causing the communication error.

Below are the exact arguments I pasted into "Terminal" that worked for me:


`/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude -C/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/etc/avrdude.conf -v -V -patmega328p -cavrisp -P/dev/cu.usbserial-14110 -b19200 -e -Ulock:w:0x3F:m`

`/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude -C/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/etc/avrdude.conf -v -V -patmega328p -cavrisp -P/dev/cu.usbserial-14110 -b19200 -Ulfuse:w:0xFF:m -Uhfuse:w:0xD7:m -Uefuse:w:0xFF:m`

`/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude -C/Users/daddy/Library/Arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/etc/avrdude.conf -v -V -patmega328p -cavrisp -P/dev/cu.usbserial-14110 -b19200 -D -Uflash:w:/private/var/folders/gb/fqk5qm6d5s9057zzjf_vhc5w0000gn/T/arduino/sketches/firmware.hex:i -Ueeprom:w:/private/var/folders/gb/fqk5qm6d5s9057zzjf_vhc5w0000gn/T/arduino/sketches/settings.hex:i`

```
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.02s

avrdude: Device signature = 0x1e950f (probably m328p)
avrdude: safemode: lfuse reads as FF
avrdude: safemode: hfuse reads as D7
avrdude: safemode: efuse reads as FF
avrdude: reading input file "/private/var/folders/gb/fqk5qm6d5s9057zzjf_vhc5w0000gn/T/arduino/sketches/firmware.hex"
avrdude: writing flash (32126 bytes):

Writing | ################################################## | 100% 37.67s

avrdude: 32126 bytes of flash written
avrdude: reading input file "/private/var/folders/gb/fqk5qm6d5s9057zzjf_vhc5w0000gn/T/arduino/sketches/settings.hex"
avrdude: writing eeprom (1024 bytes):

Writing | ################################################## | 100% 50.23s

avrdude: 1024 bytes of eeprom written

avrdude: safemode: lfuse reads as FF
avrdude: safemode: hfuse reads as D7
avrdude: safemode: efuse reads as FF
avrdude: safemode: Fuses OK (E:FF, H:D7, L:FF)

avrdude done.  Thank you
```

**Comment by ic-racer** [comment]
I'm always interested in backtracking the error. So, when I used the verbose output of Arduino IDE to get the (Path to avrdude) and (Path to config file) and (Comm Port) I inadvertently also copied the baud rate of 57600.
